### DISCUSSION TITLE
MonitorControl on Apple Silicon

### DISCUSSION THREAD

**Discussion by JoniVR** [MonitorControl/MonitorControl#379] [General]
You can use this discussion to discuss workarounds, technical info, questions etc regarding MonitorControl for Apple Silicon.
If you want to know the technical details so far, I encourage you to read #323.

**Comment by phr85** [comment]
To adjust volume of my External Philips Brilliance 499P, I'm using this App:
https://eqmac.app

**Comment by PJRomeo** [comment]
I have the Mac Mini M1 and using the Dell S2731Q monitor, getting "no supported displays found" hope to be able to use my keyboard as a volume control soon.

**Comment by stilphone** [comment]
I see in the original thread that it would be helpful to file a feedback request. I’m not a coder so I’m not 100% positive on what exactly needs fixing and how - would it be possible for someone to layout what we need to specifically request Apple to fix - i figure the more feedback requests the quicker this will get resolved for all of us..

Thanks

**Comment by tomwoerner** [comment]
I'm just gonna buy a $6000 Pro Display XDR with a stand so I can get brightness controls back

**Comment by mosherc** [comment]
I will also second that eqMac works, although it was a super buggy install for me, I had to install it twice to get it working. Make sure you turn on launch on startup. Works on LG 34BK95C-W.

**Reply by jaezh** [reply]
SoundSource is way better

**Reply by hstriepe** [reply]
SoundSource has a much wider range of functionality. It does require the installation of its Audio plug-in, which requires loosening the security on M1.
I use ALL the Rogue Amoeba tools and love them - essential!

**Comment by sergeyfilimonov** [comment]
12.2.1 — still doesn't work. Just to keep track on the macOS updates.

**Reply by jordanwan** [reply]
Of course. He means 11.2.1. Just in case anyone is searching.

**Comment by RVSarch** [comment]
I've found a workable alternative using QuickShade 2.0.3 from the AppStore to dim a Lenovo 4k monitor connected to my M1 Mac Mini via HDMI. It's a menu item, but it lets you assign shortcuts. While it will not map to the F1 and F2 keys, I have an extended keyboard and am using the F18 and F19 keys to adjust the brightness of the display. Just make sure the brightness setting on the monitor is at maximum since QuickShade does not provide direct control of the monitor. Not as good as MonitorControl but I'll use it until the bug is fixed at Apple's end.

**Reply by stepir** [reply]
with karabiner elements you can map f1/f2

wish there is a similar solution for audio - eqmac is just garbage

**Reply by phr85** [reply]
@stepir  Maybe try SpeakerAmp? https://apps.apple.com/ch/app/speakeramp-system-audio-boost/id1496955576?l=en&mt=12

**Reply by stepir** [reply]
thanks a lot will give it a try - seems to be working just fine!!!

**Comment by koenpunt** [comment]
Apart from it not working, the project also doesn't seem to compile because the OSD framework is missing the arm64 architecture. Is there a solution for that? E.g. is the OSD source available somewhere?

Edit: it seems that this private Apple framework is no longer available on Big Sur.

**Comment by hstriepe** [comment]
Soundsource by Rogue Amoeba is my goto. 
https://rogueamoeba.com/soundsource/

Cheers,
-- Harald
+ ❤️

**Comment by dch09** [comment]
Does MonitorControl still work for changing the brightness on M1 Macs? I see you guys here talking about the volume control only.

**Reply by stepir** [reply]
I don't think so - I use QuickShade + Karabiner per f1/f2 mapping

**Comment by xarbit** [comment]
Has anyone tried DisplayBuddy?

https://displaybuddy.app/

**Reply by koenpunt** [reply]
Doesn't seem to be a valid alternative;

> Yes, M1 Macs are supported using software emulation, which means DisplayBuddy modifies the display output from your Mac instead of directly changing the monitor's settings. Direct control via DDC/CI is still not supported - Apple M1's GPU doesn't expose the necessary interfaces required for that yet.

**Reply by xarbit** [reply]
Sure, that's right, I read that.. but I mean all the apps that somehow work on the M1, such as those already mentioned here must be software emulated if the native interface is missing. Or?

**Reply by yusuftrn** [reply]
i found (https://github.com/joshjon/nocturnal) nocturnal as opensource alternative.
you can try this also.
i know there is says "Apple Silicon Macs are not currently supported." but working on my mac mini and 2 external monitors.

**Comment by hstriepe** [comment]
They change brightness on the GPU, not the backlight.

**Reply by cassis35000** [reply]
Is it really an issue for the hardware ?

**Comment by waydabber** [comment]
Unfortunatelly macOS 12 Monterey still does not fix the issue it seems. :(

**Comment by xarbit** [comment]
Today marcan42 posted a dcp log on Twitter. The communications between macOS and the M1 display controller.

This design choices from Apple is interesting. If DDC is possible or not is still yet to be known. But if Apple really left DDC out of the firmware, it would impact other operating systems like Linux too. 

https://twitter.com/marcan42/status/1407928462111776768?s=21

**Comment by waydabber** [comment]
Yay! Problem solved! :)

https://github.com/MonitorControl/MonitorControl/issues/459

**Comment by PJRomeo** [comment]
There is a solution??!

**Comment by waydabber** [comment]
Yes, there is. See these:

https://gist.github.com/tao-j/c7a3c1ec7c2f59ebe45d161d620d9169
https://gist.github.com/alin23/531151c49e013554e6ca2186cef3fa90

Both volume and brightness control works perfectly (tested it on my M1 MBA and LG 4K over USB-C).

Thanks to https://github.com/tao-j and others.

**Comment by PJRomeo** [comment]
Just made my week! Thank you! And obviously the creators!

**Comment by hstriepe** [comment]
So this has not been wrapped into an app yet? Looks command line to me.

Cheers,
-- Harald
+ ❤️

**Reply by waydabber** [reply]
Not yet. But it's quite simple to modify alin23's solution (which contains the relevant parts to read the current brightness level) a bit for basic use. I did that so I can increase and decrease the current brightness by a percentage given as an argument. Then I simply added two keyboard shortcuts using iCanHazShortcut to increase and decrease brightness by +/- 5%. It's basicaly the same as MonitorControl but without the nice OSD and menu.

Hopefully super soon MonitorControl will be updated as well. :)

**Comment by PJRomeo** [comment]
Yeah, I'm too stupid to do this lmao. Im asking a friend if they can help implement it. Is there a tutorial of some sort?

**Reply by waydabber** [reply]
Ok, you don't have to do this, here is the quick&dirty thing I made:

https://gist.github.com/waydabber/ca5c4f52183598fe95cd8045adfc7623

Download the gist (by pressing the "Download ZIP" button on the top right corner), unzip it, then enter this in the terminal where you unzipped the file:

`clang -fmodules -o chbrt chbrt.m`

This will create a command line app called "chbrt". Then in the same directory in the terminal enter:

`./chbrt 10`

to increase the brightness by 10, or

`./chbrt -10`

to decrease it by 10. You can use any number between -100 and 100.

You can use Automator or your favorite app (mine is https://github.com/deseven/iCanHazShortcut ) to make a shortcut for it for easy use.

This is how I made my function keys to change the brightness using iCanHazShortcuts by binding the F1 and F2 buttons to the terminal commands explained above:

<img width="1106" alt="Screen Shot 2021-07-03 at 2 16 13" src="https://user-images.githubusercontent.com/37590873/124337814-c15ab980-dba4-11eb-8d31-05d81d8ea984.png">

Don't mind the F10-F12 shortcuts those are for controlling my Yamaha AV receiver's volume from my Mac keyboard, see the F1 and F2 buttons for brightness control.

**Reply by WolfieNuke** [reply]
Just dropping by to check on the status of M1 support for MonitorControl...

@waydabber I downloaded your code and after trying to run it, it took my external monitor (38" Dell Ultrawide via DP from M1 Mac mini) to minimum brightness and would not increase it. After a few short minutes of troubleshooting, it seems that the commands that read the current and maximum values are returning 0 for my system. As I don't know much of this API, I just overwrote the maxValue to 100 and removed the usage of the current value and the brightness control seemed to work as expected.

This is huge news, as I don't see why this cannot be easily added to MonitorControl. I'm hoping the devs are already working to implement these DDC calls for M1. Maybe soon I can get rid of my workarounds...

Thanks!

**Reply by waydabber** [reply]
Hey @WolfieNuke 

glad it worked out. You can use @tao-j's original solution as well which does not read anything but sets an absolute value:

https://gist.github.com/tao-j/c7a3c1ec7c2f59ebe45d161d620d9169

Also, check yout the latest Lunar betas which contain support for M1 Macs:

https://github.com/alin23/Lunar/issues/210

You might find links to more recent beta verisons over at Lunar's discord channel.

**Comment by tcalexwang** [comment]
I just write a draft Alfred workflow so that I can control the brightness by `br $number`

https://drive.google.com/file/d/1qP-80rTpqDWUVHnqsPixudBnNpVWwN5P/view?usp=sharing

Thanks to @waydabber, https://github.com/tao-j, and all the exciting people here.

**Comment by esusatyo** [comment]
I dived into the source code and was about to modify something in this repo itself, but I think the more correct solution is to put the fix in https://github.com/kfix/ddcctl, since this project is using that as a Swift package manager dependency. I might contribute to that repo and bump the DDC version here if I get some time.

**Reply by JoniVR** [reply]
We're using:
https://github.com/reitermarkus/DDC.swift

I have been following up on the threads but am still on vacation. 

Great to see a possible solution. I'll try to look into it later.

**Comment by waydabber** [comment]
More info about some limitations from @alin23 (see https://github.com/alin23/Lunar/issues/210 and the Lunar discord channel discussions).

- It seems M1 Mac minis cannot write DDC via HDMI (just like the issue with Intel Mini 2018). You have to use DisplayPort over USB-C. Note that some USB-C to HDMI adapters also use DisplayPort Alt Mode internally which seems to work. This is a blow since afaik M1 Minis support dual display configs only via one display being attached through HDMI.
- When the display is mirrored NSScreen.screens do not properly list all attached displays on M1 Mac Minis so if an app relies on this list to determine which monitors to control then it will have problems (but there are other methods, for example https://gist.github.com/alin23/b476a02a8cd298436848e28476aed07b can enumerate all attached external displays and attempt to control them even when displays are mirrored).

**Reply by alin23** [reply]
I just found out about [CGGetOnlineDisplayList](https://developer.apple.com/documentation/coregraphics/1454964-cggetonlinedisplaylist) which should also include mirrored displays and would fix the second point.

I'm in process of rewriting Lunar to use that instead of `NSScreen.screens` but it would need a lot of testing to make sure it doesn't break functionality for other users.

MonitorControl has a much leaner codebase (which I really envy now 😅) and should be able to replace `NSScreen.screens` more easily.

Note: don't use `CGDisplayIsOnline`, it can deadlock without a known cause.

**Reply by waydabber** [reply]
Wow, amazing! Good news! :)

**Comment by waydabber** [comment]
Here is an experimental fork of MonitorControl that works on M1 (for the impatient until DDC.swift is updated):

https://github.com/waydabber/MonitorControl

It is not fully finished yet but the app should be functional on most M1 configs with one external display attached via USB-C.

**Reply by askpete** [reply]
Hey this worked for me on a Samsung C34H892wge. 

One pretty significant bug though! 80% on the slider was 100% brightness on the monitor and when I set the slider all the way to the right the monitor blanked and I couldn't get it to go back on! The brightness must be stored so as soon as the monitor started up it went blank after 1 second. Bloody nightmare, spent over an hour trying to get it back - samsung support wanted me to get the board replaced! Eventually got the OSD up by wiggling the jog button in everydirection and the brightness was set to 125%. Reset to factory and working again thankfully .

Anyway, still awesome that you got it working at all just a warning to anyone with this monitor to be careful! I'm going to wait until the next rev after this experience though :)

**Reply by waydabber** [reply]
Huhh, that's serious.

The current version should read DDC max values but there are no safeguards implemented yet, checksums are not checked and everything is set to happen rather quickly during reads and writes. This fits the display I am testing with (an LG 27" UHD) but seems to break the C34H892.

So yes, beware! I'll improve the situation though as soon as I have the time. Still, the best option would be to wait for an official version which includes an updated DDC.Swift. :)

(ps. I added a warning and a link to your experience in the readme section.)

**Reply by alin23** [reply]
In my experience it's best to just default the max DDC value to 100 and not read it at all. Reading fails too often on M1 to be usable. 

99% of the monitors have 100 as the max so it's safest to go with that.

**Reply by waydabber** [reply]
Yes, you are right! I increased wait times and capped read values at 100 so this should be safer. I'll add checksum checks and retries in case of failure when I get to it and better handling of erroneous reads/writes in general (it will be easy since currently there is no such thing :)).

In the current state I just wanted to see if I could make everything work in pure Swift. I mostly spent about half a day trying to figure out how to implement @tao-j's original C code (especially how to get that damned IOAVServiceCreate to work as `let avService = IOAVServiceCreate(kCFAllocatorDefault) as IOAVService` did not work for whatever reason and I had to add the super obscure `?.takeRetainedValue()` to it). Also initially I did everything with pointer operations using `UnsafeMutableRawPointer` but then I reverted to using a simple array.

**Reply by waydabber** [reply]
@alin23 - also during my trials and errors i found that IOAVServiceCreate works harmoniously in Swift only if you change `typedef IOAVServiceRef` to `typedef IOAVService` in your header. I checked out how Lunar does things and as I see you use mixed C code for these operations while partially moved other stuff over to Swift - I guess that's why? :)

**Reply by alin23** [reply]
I kept the IOAVService part in C because this is time sensitive code and I don't want Swift to introduce memory management code where it is not needed. 

But it's good to know that removing Ref, ports it seamlessly to Swift.

**Comment by waydabber** [comment]
I did update the version with some initial DDC communication checks, proper checksum validations during communication and some other safeguards. Also rewritten existing m1 related stuff so now everything is even better than ever before...:

https://github.com/waydabber/MonitorControl

This should be fairly safe now. I tested it on an M1 Mini and M1 MBA with LG UHD and Lenovo displays. 

@askpete - it would be great to know if this works on your Samsung (if you are willing to take some risks).

Internal display brightness control is still broken and multi monitor support (this applies to the Mac Mini only) is still lacking. I'll look into these next time.

**Reply by askpete** [reply]
I tentatively tested it. The slider started at about 80 and I was able to make the OSD report 108 [screenshot](https://user-images.githubusercontent.com/1407347/126571830-bf879271-d52b-4cac-8f8e-c0ef96f16703.jpg)  normally only goes to 100 as you would expect. It didn't die though and I carefully stepped away after yesterday's shenanigans ;)

**Reply by waydabber** [reply]
Hey @askpete. Interesting. it seems like the display does not report the max value correctly - it probably reports 0. MonitorControl is set up in a way that it assumes 100 in this case (this is not M1 specific and is correct for like the 99% of displays as @alin23 pointed out). What I did was to set up a max cap of 100 after you reported your issue (which solves the case when the display overreports Max - like if it reports 125 instead of 100) but does not solve the case when it does not report it at all (or reports 0) and the max is actually lower than the 100 assumed in this case.

To fix this we could add an option somewhere to manually override the max DDC value range. But since you have the source code it would be interesting if you could try fixing yourself by changing the reported max from 0 to 80 for testing purposes. In order to do that you need to open the MonitorControl/Model/ExternalDisplay.swift in the XCode project and search (Alt+F) for `let max = min(UInt16(reply[7]),100)`. Then simply change the line to `let max = 80` and see what happens.

Btw, did this display work with an Intel mac before on using the Intel version of MonitorControl? If yes, then it seems the M1 version still can't read the value properly (even though now it does several retries). Maybe it's because I implemented a checksum check for read which might be too harsh for some displays.

Thank you!

**Reply by askpete** [reply]
So I just ran it inside xcode (before making changes) and noticed this on the console. 

`2021-07-22 17:53:57.323959+1000 MonitorControl[7033:224383] C34H89x (DDC.DDC.Command.contrast):
2021-07-22 17:53:57.323973+1000 MonitorControl[7033:224383]  - current ddc value: 110 - from display? true
2021-07-22 17:53:57.323978+1000 MonitorControl[7033:224383]  - maximum ddc value: 128 - from display? true`

The current value of 110 didn't change after quiting changing brightness manually and restarting. (Not that I really know what ddc is ;))

Have you committed that change? The latest rev in git is [16 June](https://github.com/MonitorControl/MonitorControl/blob/master/MonitorControl/Model/ExternalDisplay.swift)

Yeah it did work with a mbp, but I got the monitor while waiting for the mini so it was only a couple of weeks.

**Reply by waydabber** [reply]
Hmm. You seem to be looking at the wrong repository (MonitorControl/MonitorControl instead of waydabber/MonitorControl which is the experimental version). Check this one out (line 304):

https://github.com/waydabber/MonitorControl/blob/5688ced1575e608a4b6c1501563dcfa60705e75b/MonitorControl/Model/ExternalDisplay.swift#L304

But if current of max values of 110 and 128 is reported then the issue is different altogether since both max and current reporting is capped at 100 in the latest version (are you sure you pulled the latest version)?

**Reply by askpete** [reply]
my bad, pulled in wrong dir.

Display shows up in preferences but reports "No supported displays found".

**Reply by waydabber** [reply]
Yep, I added some stricter measures to check for dcc capability. Change [this](https://github.com/waydabber/MonitorControl/blob/02aac5f396221dd48013905fcf527e8baed238a4/MonitorControl/Model/ExternalDisplay.swift#L51) line from

`if m1ddcComm(send: &send, reply: &reply) {`

to

`if true {`

That will force-enable DDC even if the identification request does not produce a reply with correct checksums. The Intel version does not take this check too seriously so probably I should remove this from the m1 code as well.

**Reply by waydabber** [reply]
@askpete - what I did not mention is the fact that MonitorControl stores the min and max values in a persistent preferences store from the previous session which is preserved between app restarts and survives even when you rebuild the app after a code change. This functionality kicks in when during app start/display detection the DDC values can't be read. It's not clear what is the use of storing the max value (if no ddc max values can be read, some default max values should be assumed, there is no use of restoring them from an imaginary past DDC read success - as for the default values storing them makes sense because this allows the app to resume after a restart from the last stored values which are probably right if the user did not change them via the display's direct controls). Since I added additional checks on DDC reads (which seems to fail with your display), the app probably always reads the previously stored max values (which were incorrect in your case) - so things might seem permanently broken for you no matter what you do. This probably causes some added frustration in your case.

I will create an updated build which does not insist on a successful initial DDC read check to enable DDC (so it's always force-enabled) and does not read the max value from the persistent store if DDC reads fail but assumes a default value of 100 for brightness and volume. To make things better MonitorControl should probably add some advanced options where the user blessed with a misbehaving display can define the DDC value range accepted by his/her display.

**Reply by waydabber** [reply]
@askpete - I pushed the updated version - I know you tested a lot but if you can check this one out as well, i'd be glad to know if things work better now (you seem to have a nice non standards compliant display here which is good for testing - sadly I am surrounded with compliant displays).

If it turns out that your display still requires a lower than 100 default max level after all, please change the number from 100 to 80 (or whichever works) [here](https://github.com/waydabber/MonitorControl/blob/273d746ff96a1784b93fec3b7e8b269623b52cd1/MonitorControl/Model/ExternalDisplay.swift#L15). After you rebuild, the restored current and max values will not exceed this value no matter what was stored before in preferences.

(ps. updated the permalink, sorry)

**Reply by askpete** [reply]
Ok, not sure what's going on here. 

When I run from master sliders are disabled. Log below:
`2021-07-26 11:15:27.229353+1000 MonitorControl[21754:1311731] [plugin] AddInstanceForFactory: No factory registered for id <CFUUID 0x600003e4b1a0> F8BB1C28-BAE8-11D6-9C31-00039315CD46
2021-07-26 11:15:27.270917+1000 MonitorControl[21754:1311731] Reading Accessibility privileges - Current access status true
2021-07-26 11:15:27.272768+1000 MonitorControl[21754:1311731] Supported display found: C34H89x (Vendor: 19501, Model: 3621)
2021-07-26 11:15:27.274789+1000 MonitorControl[21754:1311731] Polling 10 times
2021-07-26 11:15:27.935300+1000 MonitorControl[21754:1311731] DDC read was unsuccessful.
2021-07-26 11:15:27.937835+1000 MonitorControl[21754:1311731] C34H89x (DDC.DDC.Command.audioSpeakerVolume):
2021-07-26 11:15:27.937866+1000 MonitorControl[21754:1311731]  - current ddc value: 42 - from display? false
2021-07-26 11:15:27.937877+1000 MonitorControl[21754:1311731]  - maximum ddc value: 0 - from display? false
2021-07-26 11:15:27.937891+1000 MonitorControl[21754:1311731] Polling 10 times
2021-07-26 11:15:27.937905+1000 MonitorControl[21754:1311731] C34H89x (DDC.DDC.Command.audioMuteScreenBlank):
2021-07-26 11:15:28.599441+1000 MonitorControl[21754:1311731] DDC read was unsuccessful.
2021-07-26 11:15:28.599519+1000 MonitorControl[21754:1311731]  - current ddc value: unknown
2021-07-26 11:15:28.599670+1000 MonitorControl[21754:1311731]  - stored maximum ddc value: 100
2021-07-26 11:15:28.610017+1000 MonitorControl[21754:1311731] Polling 10 times
2021-07-26 11:15:29.276501+1000 MonitorControl[21754:1311731] DDC read was unsuccessful.
2021-07-26 11:15:29.278171+1000 MonitorControl[21754:1311731] C34H89x (DDC.DDC.Command.luminance):
2021-07-26 11:15:29.278219+1000 MonitorControl[21754:1311731]  - current ddc value: 58 - from display? false
2021-07-26 11:15:29.278341+1000 MonitorControl[21754:1311731]  - maximum ddc value: 0 - from display? false
2021-07-26 11:15:29.286370+1000 MonitorControl[21754:1311731] Polling 10 times
2021-07-26 11:15:29.955534+1000 MonitorControl[21754:1311731] DDC read was unsuccessful.
2021-07-26 11:15:29.956683+1000 MonitorControl[21754:1311731] C34H89x (DDC.DDC.Command.contrast):
2021-07-26 11:15:29.956724+1000 MonitorControl[21754:1311731]  - current ddc value: 50 - from display? false
2021-07-26 11:15:29.956793+1000 MonitorControl[21754:1311731]  - maximum ddc value: 0 - from display? false
2021-07-26 11:15:49.788884+1000 MonitorControl[21754:1311731] Metal API Validation Enabled
2021-07-26 11:15:49.823862+1000 MonitorControl[21754:1311731] +[MTLIOAccelDevice registerDevices]: Zero Metal services found
`
When I force max and current to be 100 the slider works but has no effect. 

So thinking about it, since (my second attempt was void as I hadn't actually pulled your changes) it hasn't worked since the first attempt.

**Reply by waydabber** [reply]
@askpete , thanks for the report, I'll look into it!

Let's move the conversation to the following thread since it is now unlocked. As soon as there is an update I'll let you know there!

[MonitorControl not working on Apple Silicon (M1) Macs #323](https://github.com/MonitorControl/MonitorControl/issues/323)

Future code updates will be in the main repository's `experimental/apple-silicon` branch which can be accessed here:

https://github.com/MonitorControl/MonitorControl/tree/experimental/apple-silicon

**Comment by waydabber** [comment]
All right, M1 built in display brightness control is now fixed as well. Brightness control moved to the DisplayServices private framework it seems and works a little bit differently, that's why it was broken on M1 Macs. I tested the updated solution on my M1 MBA and works fine,  I suppose it works the same on the M1 MBP and iMac.

https://github.com/waydabber/MonitorControl

Soon I'll be running out of stuff to do... :)

@JoniVR, I understand that the Apple Silicon thread is locked again in issues but it would be useful that all those guys who wanted to testing would be able to test something (if you approve of course).  :)

**Reply by JoniVR** [reply]
Hi @waydabber thanks for reaching out and putting in some effort, if you want, we can create a separate branch and update post on the main repo and issue, if it proves stable enough we can already release a pre-release version. It’s a bit late here to do it now but I can try to set it all up later this week?

**Reply by waydabber** [reply]
That would be great! Later this week is perfect, no need to rush, I still need to finish up things anyway. Thank you!

**Reply by hirenr** [reply]
Does volume work?

**Reply by waydabber** [reply]
Yes, it should work (works fine on the displays I tested).

**Reply by iswitch** [reply]
Great! LG 27UK850 and MacBook Pro works!

**Comment by hstriepe** [comment]
Still does not find a Dell U3818DW unless already plugged in.
Controls the Brightness and Contrast.
Keys do not show HUD.

My normal use case is:
- Monitor Control always runs
- Activates and works on the USB-C monitor, when plugged in.
Cheers,
-- Harald
+ ❤️

**Reply by waydabber** [reply]
Hey @hstriepe, thanks for the report!

- Can you tell me the type of the Mac and the OS version?
- What happens if you 1) plug in, 2) unplug and then 3) plug your display back in and 4) open and then 5) close the lid while the display is plugged in? Does the display list change accordingly in Preferences/Display at each step? Can the internal and external display (whichever is actvie) be controlled properly after each step (don't forget that you first have to click with your mouse on the display you want to control via the keyboard).
- The OSD thing is interesting. Can you please try this: [showosd](https://github.com/waydabber/showosd)? If you complie it and enter `showosd brightness 40` does a brightness OSD appear at 40%?

**Comment by waydabber** [comment]
Well, it seems I hit a wall. Proper multi monitor support (for external displays, this does not concern the internal+external use case) is simply too difficult to achieve (it needs a rather convoluted service matching process with all kinds of caveats). I could do it in a haphazard way but I don't want to do that and I just dig into deeper and deeper holes trying to figure out how to do this the right way (I think @alin23 could talk about that a lot - I checked out his code in Lunar trying to solve this and my brain fried :)). So I'll just leave it at that.

This only affects M1 Mac minis as of now (all other M1 Macs support a single external display) when a HDMI is connected but since DDC doesn't seem work on M1 mini's HDMI port anyway such an user wouldn't really use MonitorControl even if this thing was fixed (but if a Mac mini multi monitor user wants to control at least the display attached through USB-C/DP, that will work or if not there is a workaround which I described [here](https://github.com/waydabber/MonitorControl/blob/master/README.md)).

The issue will have to be resolved though as later Apple Silicon machines will surely support more than one external display.

**Comment by hstriepe** [comment]
Seems to work ok now - build 1169
> Can you tell me the type of the Mac and the OS version?
PowerBook Pro M1, Big Sur 11.5
> What happens if you 1) plug in, 2) unplug and then 3) plug your display back in and 4) open and then 5) close the lid while the display is plugged in? Does the display list change accordingly in Preferences/Display at each step? Can the internal and external display (whichever is actvie) be controlled properly after each step (don't forget that you first have to click with your mouse on the display you want to control via the keyboard).
Replugging works
> The OSD thing is interesting. Can you please try this: showosd <https://github.com/waydabber/showosd>? If you complie it and enter showosd brightness 40 does a brightness OSD appear at 40%?
OSD now working

**Reply by waydabber** [reply]
Amazing, thanks for the report!

**Comment by waydabber** [comment]
The new v3.0.0 version has been released with Apple Silicon support. I will lock this discussion - if you have any problems with the release, please open an Issue or let us know in the release discussion. Thank you!
